### PR TITLE
LLL: implement the UNSET and WITH keywords

### PR DESCRIPTION
**Pull request by axic**
Part of #1288.

### CODE REVIEW

**Comment by axic** [comment]
Don't merge it yet, this might use some refactoring.

**Comment by axic** [comment]
cc @benjaminion @zigguratt

**Comment by zigguratt** [comment]
@axic That's great! `with` is one of those keywords that was in Serpent's LLL but missing in Solidity's LLL. Excellent work.

**Comment by benjaminion** [comment]
Looks good to me. I tested with nested "with"s  and that works as well :-)

**Comment by axic** [comment]
@zigguratt @benjaminion should we have this?

**Comment by benjaminion** [comment]
> @benjaminion should we have this?

If `with` behaved as described in #1288 (using the stack to hold the value) then I'd be a big fan. However, it looks to be implemented here as an in-memory variable. I don't really like variables for two main reasons:

 1. They interfere with memory management in ways that are hard to handle nicely.
 2. The memory space isn't reclaimed after a variable is undef'd (or goes out of scope in this case).

Having said that, others may find a use for this, so I wouldn't be against it, and there is a bigger picture to consider.

**Comment by axic** [comment]
@benjaminion I agree it would be better keeping it on the stack - this is a quick and dirty version to have the feature for compatibility. Would be happy to have a stack version if you have time to work on it.

**Comment by pirapira** [comment]
Is @benjaminion still working on this, or should we try to merge this version?

**Comment by benjaminion** [comment]
Hi @pirapira, @axic - not working on this currently. I did some initial analysis, but as far as I can see, persuading the LLL compiler to make better use of the stack would require a big re-organisation.  It would be great to do sometime - I've got use cases in mind around supporting proper functions in LLL, and integration with Solidity as an alternative to ASM - but it's beyond my powers at this time.

Happy for this to be merged.

File: liblll/CodeFragment.cpp [hunk 1]
```diff
@@ -259,6 +259,7 @@ void CodeFragment::constructOperation(sp::utree const& _t, CompilerState& _s)
 		}
 		else if (us == "SET")
 		{
+			// TODO: move this to be a stack variable (and not a memory variable)
```
Comment (chriseth) [inline liblll/CodeFragment.cpp:262]: Wouldn't that change the semantics of the language?

File: liblll/CodeFragment.cpp [hunk 2]
```diff
@@ -268,13 +269,51 @@ void CodeFragment::constructOperation(sp::utree const& _t, CompilerState& _s)
 			m_asm.append((u256)varAddress(firstAsString(), true));
 			m_asm.append(Instruction::MSTORE);
 		}
+		else if (us == "UNSET")
+		{
+			// TODO: this doesn't actually free up anything, since it is a memory variable (see "SET")
+			if (_t.size() != 2)
+				error<IncorrectParameterCount>();
+			auto it = _s.vars.find(firstAsString());
+			if (it != _s.vars.end())
```
Comment (chriseth) [inline liblll/CodeFragment.cpp:278]: else no error? Then why do we have "unset" at all?

File: liblll/CodeFragment.cpp [hunk 3]
```diff
@@ -268,13 +269,51 @@ void CodeFragment::constructOperation(sp::utree const& _t, CompilerState& _s)
...
+		else if (us == "WITH")
+		{
+			if (_t.size() != 4)
+				error<IncorrectParameterCount>();
+			string key = firstAsString();
+			if (_s.vars.find(key) != _s.vars.end())
+				error<InvalidName>(string("Symbol already used: ") + key);
+
+			// Create variable
+			// TODO: move this to be a stack variable (and not a memory variable)
```
Comment (chriseth) [inline liblll/CodeFragment.cpp:297]: Isn't this a vital requirement?

File: liblll/CodeFragment.cpp [hunk 4]
```diff
@@ -268,13 +269,51 @@ void CodeFragment::constructOperation(sp::utree const& _t, CompilerState& _s)
...
+			size_t c = 0;
+			for (auto const& i: _t)
+				if (c++ == 2)
```
Comment (chriseth) [inline liblll/CodeFragment.cpp:300]: Please use `++c` here

File: liblll/CodeFragment.cpp [hunk 5]
```diff
@@ -268,13 +269,51 @@ void CodeFragment::constructOperation(sp::utree const& _t, CompilerState& _s)
...
+			if (_t.size() != 4)
+				error<IncorrectParameterCount>();
+			string key = firstAsString();
+			if (_s.vars.find(key) != _s.vars.end())
+				error<InvalidName>(string("Symbol already used: ") + key);
+
+			// Create variable
+			// TODO: move this to be a stack variable (and not a memory variable)
+			size_t c = 0;
+			for (auto const& i: _t)
```
Comment (chriseth) [inline liblll/CodeFragment.cpp:299]: Please use a different iterator.

**Review by chriseth — APPROVED** [review]
Haven't fully checked it, but seems to be in line with lll's code quality.

**Comment by axic** [comment]
> Haven't fully checked it, but seems to be in line with lll's code quality.

Yep, it is a copy of the horrible style it has.

**Pull request merged**